plugin/pkg/shoot/managedseed/admission.go

```diff
@@ -0,0 +1,179 @@
+
+// New creates a new ManagedSeed admission plugin.
+func New() (*ManagedSeed, error) {
+	return &ManagedSeed{
+		Handler: admission.NewHandler(admission.Delete),
+	}, nil
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (v *ManagedSeed) AssignReadyFunc(f admission.ReadyFunc) {
+	v.readyFunc = f
+	v.SetReadyFunc(f)
+}
+
+// SetInternalCoreClientset sets the garden core clientset.
+func (v *ManagedSeed) SetInternalCoreClientset(c coreclientset.Interface) {
+	v.coreClient = c
+}
+
+// SetSeedManagementClientset sets the garden seedmanagement clientset.
+func (v *ManagedSeed) SetSeedManagementClientset(c seedmanagementclientset.Interface) {
+	v.seedManagementClient = c
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (v *ManagedSeed) ValidateInitialization() error {
+	if v.coreClient == nil {
+		return errors.New("missing garden core client")
+	}
+	if v.seedManagementClient == nil {
+		return errors.New("missing garden seedmanagement client")
+	}
+	return nil
+}
+
+var _ admission.ValidationInterface = &ManagedSeed{}
+
+// Validate validates if the Shoot can be deleted. If the
+func (v *ManagedSeed) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	// Wait until the caches have been synced
+	if v.readyFunc == nil {
+		v.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !v.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	// Ignore all kinds other than Shoot
+	if a.GetKind().GroupKind() != core.Kind("Shoot") {
+		return nil
+	}
+
+	// Ignore updates to status or other subresources
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
+	switch {
+	case a.GetName() == "":
+		return v.validateDeleteCollection(ctx, a)
+	default:
+		return v.validateDelete(ctx, a)
+	}
+}
+
+func (v *ManagedSeed) validateDeleteCollection(ctx context.Context, a admission.Attributes) error {
+	shoots, err := v.getShoots(ctx, labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, shoot := range shoots {
+		if err := v.validateDelete(ctx, v.newAttributesWithName(a, shoot.Name)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v *ManagedSeed) validateDelete(ctx context.Context, a admission.Attributes) error {
+	managedSeed, err := kutil.GetManagedSeed(ctx, v.seedManagementClient, a.GetNamespace(), a.GetName())
+	if err != nil {
+		return apierrors.NewInternalError(fmt.Errorf("could not get ManagedSeed for shoot '%s/%s': %v", a.GetNamespace(), a.GetName(), err))
+	}
+	if managedSeed == nil {
+		return nil
+	}
+
+	return admission.NewForbidden(a, fmt.Errorf("cannot delete shoot %s/%s since it is still referenced by a managed seed", a.GetNamespace(), a.GetName()))
+}
+
+func (d *ManagedSeed) newAttributesWithName(a admission.Attributes, name string) admission.Attributes {
```

Comment:
Receiver names are different
```suggestion
func (v *ManagedSeed) newAttributesWithName(a admission.Attributes, name string) admission.Attributes {
```

Reply:
OK

Reply:
Done